File: backtester/backend/schemas/backtest.py
"""Backtest API Schemas.

"""

from datetime import date
from typing import Any, Dict, List, Optional, Union

from pydantic import BaseModel, Field


class BacktestRequest(BaseModel):
    """백테스트 요청"""
    strategy_id: str = Field(..., description="전략 ID")
    symbols: List[str] = Field(..., description="종목 코드 리스트")
    start_date: Union[str, date] = Field(..., description="시작일 (YYYY-MM-DD)")
    end_date: Union[str, date] = Field(..., description="종료일 (YYYY-MM-DD)")
    initial_capital: float = Field(default=100_000_000, description="초기 자본")
    param_overrides: Optional[Dict[str, Any]] = Field(
        default=None,
        description="파라미터 오버라이드 (예: {'period': 21, 'oversold': 25})"
    )
    commission_rate: Optional[float] = Field(
        default=0.00015,
        description="수수료율 (기본 0.00015 = 0.015%)"
    )
    tax_rate: Optional[float] = Field(
        default=0.002,
        description="거래세율 (기본 0.002 = 0.2%, 매도 시)"
    )
    slippage: Optional[float] = Field(
        default=0.0,
        description="슬리피지 (기본 0 = 0%)"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "strategy_id": "rsi_oversold",
                "symbols": ["005930", "000660"],
                "start_date": "2024-01-01",
                "end_date": "2024-12-31",
                "initial_capital": 100000000,
                "param_overrides": {"period": 21, "oversold": 25},
                "commission_rate": 0.00015,
                "tax_rate": 0.002,
                "slippage": 0.001,
            }
        }


class BacktestResponse(BaseModel):
    """백테스트 응답"""
    success: bool = True
    data: Dict[str, Any]
    message: Optional[str] = None

File: backtester/frontend/src/types/backtest.ts
/**
 * 백테스트 타입 정의 - Lean 전체 통계 반영
 */

export interface BacktestRequest {
  strategy_id: string;
  symbols: string[];
  start_date: string;
  end_date: string;
  initial_capital: number;
  commission_rate?: number;
  tax_rate?: number;
  slippage?: number;
  param_overrides?: Record<string, number>;
}

export interface CustomBacktestRequest {
  yaml_content: string;
  symbols: string[];
  start_date: string;
  end_date: string;
  initial_capital: number;
  commission_rate?: number;
  tax_rate?: number;
  slippage?: number;
}

/**
 * Lean 전체 통계 - 카테고리별 그룹화
 */
export interface PerformanceMetrics {
  basic: {
    total_return: number;
    annual_return: number;
    max_drawdown: number;
    start_equity: number;
    end_equity: number;
  };
  risk: {
    sharpe_ratio: number;
    sortino_ratio: number;
    probabilistic_sharpe: number;
  };
  greeks: {
    alpha: number;
    beta: number;
  };
  volatility: {
    annual_std_dev: number;
    annual_variance: number;
  };
  benchmark: {
    information_ratio: number;
    tracking_error: number;
    treynor_ratio: number;
  };
  trading: {
    total_orders: number;
    win_rate: number;
    loss_rate: number;
    avg_win: number;
    avg_loss: number;
    profit_loss_ratio: number;
    expectancy: number;
  };
  other: {
    total_fees: number;
    portfolio_turnover: number;
    estimated_capacity: number;
    drawdown_recovery: number;
  };
}

export interface TradeInfo {
  symbol: string;
  direction: "Buy" | "Sell";
  quantity: number;
  price: number;
  time: string;
  profit?: number;
  profit_percent?: number;
}

export interface BacktestResult {
  run_id: string;
  strategy_name: string;
  start_date: string;
  end_date: string;
  initial_capital: number;
  final_capital: number;
  net_profit: number;
  net_profit_percent: number;
  metrics: PerformanceMetrics;
  equity_curve: Record<string, number>;
  benchmark_curve?: Record<string, number>;  // KOSPI 수익률 커브 (%)
  trades_count: number;
  trades?: TradeInfo[];  // 거래 내역
}

export interface BacktestResponse {
  success: boolean;
  data: BacktestResult;
  message?: string;
}

/**
 * 차트 데이터 포인트
 */
export interface EquityDataPoint {
  date: string;
  value: number;
}

export interface DrawdownDataPoint {
  date: string;
  drawdown: number;
}
